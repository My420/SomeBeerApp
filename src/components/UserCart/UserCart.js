import React from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import styles from './UserCart.module.scss';
import withCart from '../HOC/CartWrapper/CartWrapper';
import ListGroup from '../ListGroup/ListGroup';
import ListGroupItem from '../ListGroupItem/ListGroupItem';
import CartCard from '../CartCard/CartCard';
import Button from '../Button/Button';

export const UserCart = ({
  cartData,
  deleteItemFromCart,
  changeCartItemAmount
}) => {
  // eslint-disable-next-line no-console
  console.log('render ==================== userCart');

  const isEmpty = cartData.size === 0;

  const getCartList = () => {
    if (isEmpty) {
      return <ListGroupItem className={styles.empty}>is empty.</ListGroupItem>;
    }

    const data = cartData.toList().toJS();
    const list = data.map(elem => {
      return (
        <ListGroupItem className={styles.item} key={elem.item.id}>
          <CartCard
            data={elem}
            changeAmount={changeCartItemAmount}
            deleteCard={deleteItemFromCart}
          />
        </ListGroupItem>
      );
    });

    return list;
  };

  return (
    <React.Fragment>
      <p className={styles.title}>Your Shopping Cart</p>
      <ListGroup className={styles.list}>{getCartList()}</ListGroup>
      {!isEmpty ? (
        <div className={styles.controls}>
          <Button className={styles.proceed} disabled>
            Proceed to checkout
          </Button>
        </div>
      ) : null}
    </React.Fragment>
  );
};

UserCart.propTypes = {
  cartData: PropTypes.instanceOf(Map).isRequired,
  deleteItemFromCart: PropTypes.func.isRequired,
  changeCartItemAmount: PropTypes.func.isRequired
};

export default withCart(UserCart);
