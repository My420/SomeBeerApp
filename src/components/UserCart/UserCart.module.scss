@import '../../sass/variables.scss';

.list {
  list-style: none;
  margin-bottom: 20px;
}

.item {
  margin-bottom: 10px;
}

.title,
.empty {
  margin-bottom: 10px;
  font-size: 24px;
  line-height: 28px;
  text-align: center;
}

.controls {
  width: 100%;
  margin: 20px 0;
}

.proceed {
  @include link-button;
  margin: 0 auto;
}
