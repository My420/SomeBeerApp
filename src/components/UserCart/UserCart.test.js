import React from 'react';
import { Map } from 'immutable';
import { shallow } from 'enzyme';
import { UserCart } from './UserCart';

describe('test <UserCart /> component', () => {
  const changeMock = jest.fn();
  const delMock = jest.fn();

  const data = Map({
    1: {
      amount: 1,
      item: { id: 1, image_url: 'test_url1', name: 'test name 1' }
    },

    2: {
      amount: 1,
      item: { id: 2, image_url: 'test_url2', name: 'test name 2' }
    }
  });

  const emptyData = Map({});

  describe('test component when cardData is not empty', () => {
    const wrapper = shallow(
      <UserCart
        cartData={data}
        changeCartItemAmount={changeMock}
        deleteItemFromCart={delMock}
      />
    );
    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test(`should render ${data.size} CartCard components`, () => {
      const components = wrapper.find('CartCard');
      expect(components).toHaveLength(data.size);
    });

    test('should render first CartCard component with correct props', () => {
      const component = wrapper.find('CartCard').first();
      expect(component.prop('data')).toEqual(data.get('1'));
      expect(component.prop('changeAmount')).toEqual(changeMock);
      expect(component.prop('deleteCard')).toEqual(delMock);
    });

    test('should render second CartCard component with correct props', () => {
      const component = wrapper.find('CartCard').last();
      expect(component.prop('data')).toEqual(data.get('2'));
      expect(component.prop('changeAmount')).toEqual(changeMock);
      expect(component.prop('deleteCard')).toEqual(delMock);
    });

    test('should render proceed button with correct props', () => {
      const component = wrapper.find('Button');
      expect(component.prop('children')).toBe('Proceed to checkout');
      expect(component.prop('disabled')).toBeTruthy();
    });
  });

  describe('test component when cardData is empty', () => {
    const wrapper = shallow(
      <UserCart
        cartData={emptyData}
        changeCartItemAmount={changeMock}
        deleteItemFromCart={delMock}
      />
    );
    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('should not render proceed button', () => {
      const component = wrapper.find('Button');
      expect(component).toHaveLength(0);
    });

    test('should render ListGroupItem with correct props', () => {
      const component = wrapper.find('ListGroupItem');
      expect(component).toHaveLength(1);
      expect(component.prop('className')).toBe('empty');
      expect(component.prop('children')).toBe('is empty.');
    });
  });
});
