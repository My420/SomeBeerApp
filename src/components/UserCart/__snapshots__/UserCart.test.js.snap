// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <UserCart /> component test component when cardData is empty should match snapshot 1`] = `
<Fragment>
  <p
    className="title"
  >
    Your Shopping Cart
  </p>
  <ListGroup
    className="list"
    tag="ul"
  >
    <ListGroupItem
      className="empty"
    >
      is empty.
    </ListGroupItem>
  </ListGroup>
</Fragment>
`;

exports[`test <UserCart /> component test component when cardData is not empty should match snapshot 1`] = `
<Fragment>
  <p
    className="title"
  >
    Your Shopping Cart
  </p>
  <ListGroup
    className="list"
    tag="ul"
  >
    <ListGroupItem
      className="item"
      key="1"
    >
      <CartCard
        changeAmount={[MockFunction]}
        data={
          Object {
            "amount": 1,
            "item": Object {
              "id": 1,
              "image_url": "test_url1",
              "name": "test name 1",
            },
          }
        }
        deleteCard={[MockFunction]}
      />
    </ListGroupItem>
    <ListGroupItem
      className="item"
      key="2"
    >
      <CartCard
        changeAmount={[MockFunction]}
        data={
          Object {
            "amount": 1,
            "item": Object {
              "id": 2,
              "image_url": "test_url2",
              "name": "test name 2",
            },
          }
        }
        deleteCard={[MockFunction]}
      />
    </ListGroupItem>
  </ListGroup>
  <div
    className="controls"
  >
    <Button
      activeStyle=""
      className="proceed"
      disabled={true}
      onClick={[Function]}
    >
      Proceed to checkout
    </Button>
  </div>
</Fragment>
`;
